### PR TITLE
User Operations Migration

### LINKED ISSUE #6: Migrate the User APIs
Migrate the following APIs mentioned over in this [location](https://github.com/nvseshaiah2013/hisaab_api/blob/master/routes/user.js)

- [x] Base Url "/"
- [x] /login Endpoint
- [x] /singup Endpoint
- [x] /changePassword Endpoint
- [x] /forgotPassword Endpoint

### PR DESCRIPTION
**Pull request by nvseshaiah2013**
closes #6

### CODE REVIEW

File: src/main/java/com/kitaab/hisaab/ledger/entity/user/TokenMetadata.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.index.IndexDirection;
+import org.springframework.data.mongodb.core.index.Indexed;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+import java.util.Date;
+
+@Data
+@Builder(setterPrefix = "with")
+@Document(collection = "users")
```
Comment (nvseshaiah2013) [inline src/main/java/com/kitaab/hisaab/ledger/entity/user/TokenMetadata.java:16]: collection name users ?

File: src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    @Override
+    public SuccessResponse resetPassword(String authToken, String newPassword) {
+        String username = jwtService.extractUsername(authToken);
+        log.info("Trying to change password for user: {}", username);
+        var tokenMetadata = Optional.ofNullable(tokenMetadataRepository.findByUsernameOrderByCreatedAtDesc(username))
+                .orElseThrow(() -> new FlowBreakerException(ExceptionEnum.USERNAME_NOT_FOUND.getErrorCode(),
+                        ExceptionEnum.USERNAME_NOT_FOUND.getMessage(username)));
+        Boolean isValid = jwtService.validateToken(tokenMetadata.getToken(), username);
+        if (isValid && Objects.equals(tokenMetadata.getToken(), authToken)) {
```
Comment (nvseshaiah2013) [inline src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java:147]: What happens if condition is not true.

**Comment by nvseshaiah2013** [comment]
@samir937 , @rahul-ojha-07 Pls review

File: src/main/java/com/kitaab/hisaab/ledger/exception/CustomExceptionHandler.java [hunk 3]
```diff
@@ -60,4 +60,14 @@ private ResponseEntity<ErrorResponse> getErrorResponseResponseEntity(Exception e
                 traceId, spanId);
         return new ResponseEntity<>(message, HttpStatus.BAD_REQUEST);
     }
+
+    @ExceptionHandler(FlowBreakerException.class)
+    public ResponseEntity<ErrorResponse> handleApplicationException(Throwable ex) {
+        traceId = Objects.requireNonNull(tracer.currentSpan()).context().traceId();
+        spanId = Objects.requireNonNull(tracer.currentSpan()).context().spanId();
+        var exception = (FlowBreakerException) ex;
```
Comment (samir937) [inline src/main/java/com/kitaab/hisaab/ledger/exception/CustomExceptionHandler.java:68]: we can directly use FlowBreakerException ex in method parameter . Why an extra conversion ?

Comment (nvseshaiah2013) [reply src/main/java/com/kitaab/hisaab/ledger/exception/CustomExceptionHandler.java]: done the changes

File: src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                    ExceptionEnum.PASSWORD_RESET_TOKEN_INVALID);
+        }
+        return new SuccessResponse(HttpStatus.OK,
+                MessageFormat.format(ApplicationConstants.PASSWORD_RESET_SUCCESS, username));
+    }
+
+    @Override
+    public SuccessResponse getUsers(String usernamePrefix) {
+        log.info("Finding all Users starting with Username : {}", usernamePrefix);
+        List<User> users = userRepository.findByUsernameStartsWith(usernamePrefix);
```
Comment (nvseshaiah2013) [inline src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java:207]: List of users should not show the currently logged in user.

Comment (rahul-ojha-07) [reply src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java:207]: updated code for same

**Comment by rahul-ojha-07** [comment]
@nvseshaiah2013 @samir937 - ok review from my side 
approve and merge

File: src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java (hunk 4 above)
Comment (nvseshaiah2013) [reply src/main/java/com/kitaab/hisaab/ledger/service/impl/UserServiceImpl.java:207]: this code will instead keep only the current user. We can keep the condition as following `user -> !Objects.equals(user.getUsername(), userDetails.getUsername())`

**Pull request merged**